File: backend/typescript-node-express-hive/src/shared/security/password.ts
import * as argon2 from "argon2";
import { newInvalidCredentialsError } from "../errors/app-error.js";

export interface PasswordHasher {
    hash(password: string): Promise<string>;
    compare(hashedPassword: string, password: string): Promise<void>;
}

export class Argon2PasswordHasher implements PasswordHasher {
    async hash(password: string): Promise<string> {
        return await argon2.hash(password);
    }

    async compare(hashedPassword: string, password: string): Promise<void> {
        const valid = await argon2.verify(hashedPassword, password);
        if (!valid) {
            throw newInvalidCredentialsError("Invalid email or password");
        }
    }
}

File: backend/python-fastapi-vertical/app/shared/security/password.py
import bcrypt

from shared.errors import new_invalid_credentials_error


def hash_password(password: str) -> str:
    salt = bcrypt.gensalt()
    hashed = bcrypt.hashpw(password.encode("utf-8"), salt)
    return hashed.decode("utf-8")


async def compare_password(hashed_password: str, password: str) -> None:
    valid = bcrypt.checkpw(password.encode("utf-8"), hashed_password.encode("utf-8"))
    if not valid:
        raise new_invalid_credentials_error("Invalid email or password")
